File: 프로그래머스/1/172928. 공원 산책/공원 산책.swift
import Foundation

func solution(_ park:[String], _ routes:[String]) -> [Int] {
    
    var (h, w) :(Int, Int) = (0, 0)
    var count = 0
    
    // park 이중 배열로 나타내기
    var newPark: [[String]] = []
    for element in park{
        var ap: [String] = []
        ap = element.map { String($0) }
        // 현재위치 파악하기
        if let x = ap.firstIndex(of: "S"){
            ( h, w) = ( count, Int(x))
        }
        newPark.append(ap)
        count += 1
    }
    
    // 동서남북으로 이동할 배열 만들기
    let op = ["N", "S", "W", "E"]
    let move = [[-1, 0], [1, 0], [0, -1], [0, 1]]
    
    // routes 배열을 위치와 칸 수로 구분하기
    for order in routes{
        let a = (order.components(separatedBy: " ").map {$0})
        let ( currentOp, n) = (a[0], Int(a[1])!)
        let index = op.firstIndex(of: currentOp)!
        
        var (hx, wx) = (h, w)
        var c = 0
        for _ in 1...n{
            hx += (move[index][0])
            wx += (move[index][1])
            // 1. 공원 벗어나는지 확인
            // 2. 장애물 만나는지 확인
            if hx >= 0 && hx < newPark.count && wx >= 0 && wx < newPark[0].count && newPark[hx][wx] != "X"{
                c += 1
            }
            if c == n {
                ( h, w ) = ( hx, wx )
            }
        }
    }
    return [h, w]
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    
    answer = []
    
    di = [0, 1, 0, -1]
    dj = [1, 0, -1, 0]
    
    location_dict = {'E':0, 'S':1, 'W':2, 'N':3}
    
    ci, cj = 0, 0
    
    # 현재 위치 찾기
    for index, pa in enumerate(park):
        if 'S' in pa:
            ci = index
            cj = pa.index('S')
            break
    
    # 좌표 이동하기
    for route in routes:
        l, nm = route.split()
        k = location_dict[l]
        cnt = 0
        
        for n in range(1, int(nm)+1):
            
            ni = ci + di[k] * n
            nj = cj + dj[k] * n
            
            if 0 > ni or ni >= len(park) or 0 > nj or nj >= len(park[0]):
                break
            
            if park[ni][nj] == 'X':
                break
            
            cnt += 1
        
        if cnt == int(nm):
            ci, cj = ni, nj
        
    answer = [ci, cj]
    
    return answer